api/app/services/highlighter.py

```diff
@@ -102,7 +105,66 @@ def highlight_paragraph(self, query_state, para_state,
 
             new_paragraph.append(sent)
             last_pos += len(sent) + 1
-        return ' '.join(new_paragraph), highlights
+
+        highlighted = ' '.join(new_paragraph)
+        return self.adjust_highlights(original_paragraph, highlighted, highlights)
+
+    def adjust_highlights(self, original_text: str, highlighted_text: str,
+                          highlights: List[List[Tuple[int]]]):
+        """
+        Adjusts highlights based on extra spaces introduced by the tokenization process.
+        Iterates over highlighted and original text simultaneously to compute character positions.
+        """
+
+        if len(highlights) == 0:
```

Comment:
```suggestion
        if highlights:
```
This handles both `None` and `[]`